src/main/java/org/openstreetmap/atlas/geography/boundary/DynamicGridIndexBuilder.java

```diff
@@ -1,40 +1,245 @@
+     *
+     * @author mkalender
+     */
+    class GridWorkItem
+    {
+        private final double minX;
+        private final double minY;
+        private final double maxX;
+        private final double maxY;
+        private final Polygon polygon;
+
+        /**
+         * Default constructor
+         *
+         * @param minX
+         *            Lower bound on X-axis
+         * @param minY
+         *            Lower bound on Y-axis
+         * @param maxX
+         *            Upper bound on X-axis
+         * @param maxY
+         *            Upper bound on Y-axis
+         * @param polygon
+         *            Boundary in {@link Polygon} format
+         */
+        GridWorkItem(final double minX, final double minY, final double maxX, final double maxY,
+                final Polygon polygon)
+        {
+            this.minX = minX;
+            this.minY = minY;
+            this.maxX = maxX;
+            this.maxY = maxY;
+            this.polygon = polygon;
+        }
+
+        double maxX()
+        {
+            return this.maxX;
+        }
+
+        double maxY()
+        {
+            return this.maxY;
+        }
+
+        double minX()
+        {
+            return this.minX;
+        }
+
+        double minY()
+        {
+            return this.minY;
+        }
+
+        Polygon polygon()
+        {
+            return this.polygon;
+        }
+    }
+
+    /**
+     * Worker that acts like both producer and consumer. Takes an {@link GridWorkItem} from work
+     * queue and processes it.
+     *
+     * @author mkalender
+     */
+    class Processor implements Runnable
+    {
+        private final BlockingQueue<GridWorkItem> queue;
+        private final STRtree index;
+        private final STRtree rawIndex;
+
+        /**
+         * Default constructor.
+         *
+         * @param queue
+         *            {@link BlockingQueue} holding {@link GridWorkItem}s to process
+         * @param index
+         *            {@link STRtree} to be built
+         * @param rawIndex
+         *            {@link STRtree} holding raw boundaries
+         */
+        Processor(final BlockingQueue<GridWorkItem> queue, final STRtree index,
+                final STRtree rawIndex)
+        {
+            this.queue = queue;
+            this.index = index;
+            this.rawIndex = rawIndex;
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                while (!this.queue.isEmpty())
+                {
+                    final GridWorkItem itemToProcess = this.queue.poll();
+
+                    // Queue might have been emptied by another thread. That gives an null item.
+                    // Check for null item to avoid an exception.
+                    if (itemToProcess != null)
+                    {
+                        this.process(itemToProcess.minX, itemToProcess.minY, itemToProcess.maxX,
```

Comment:
How about passing the grid work item directly? Just a suggestion.

Reply:
Yeah, that'll be cleaner.